### PR TITLE
PICARD-840: Warn people who try to load very large amounts of files

### PR DESCRIPTION
**Pull request by emhazen**
Default maximum: 5000

The default maximum can be changed by the user in Options > General within the range 1 - 100,000,000.

If a user tries to load more files than the set maximum, a message box will alert the user and ask if they would like to continue.

### CODE REVIEW

File: picard/tagger.py [hunk 1]
```diff
@@ -355,20 +356,44 @@ def add_files(self, filenames, target=None):
...
+    def check_load(self, new_files):
+        #Load large amounts of files
+        if len(new_files) > config.setting["file_import_threshold"]:
+            QMessageBox = QtGui.QMessageBox
+
+            msg = QMessageBox()
+            msg.setIcon(QMessageBox.Warning)
+            msg.setWindowModality(QtCore.Qt.WindowModal)
+            msg.setWindowTitle(_(u"Large File Load"))
+            msg.setText(_(u"Are you sure you want to load %s files? Picard may "
```
Comment (zas) [inline picard/tagger.py]: `%s` should be replaced by `%d` here, value is an integer.
Also, while it is unlikely you have 0 or 1 file here, plural forms depend on languages, so you have to do something like https://github.com/metabrainz/picard/blob/1d1579130b02f9ecb496d7ccae925e77d4269ac5/picard/tagger.py#L398-L402

It will allow translators to provide proper translations.

Comment (emhazen) [inline picard/tagger.py]: I also changed the "Load Files" button to just say "Load", to keep things simple.

File: picard/ui/ui_options_general.py [hunk 2]
```diff
@@ -98,4 +118,6 @@ def retranslateUi(self, GeneralOptionsPage):
         self.groupBox_2.setTitle(_("General"))
         self.analyze_new_files.setText(_("Automatically scan all new files"))
         self.ignore_file_mbids.setText(_("Ignore MBIDs when loading new files"))
+        self.import_settings.setTitle(_("Importing"))
+        self.label_import.setText(_("Maximum number of files to import without error:"))
```
Comment (zas) [inline picard/ui/ui_options_general.py]: It isn't an error, but a warning, so i would rephrase to: "Warn if user tries to load more than this number of files at once:" (but i'm not a native english speaker).

File: picard/ui/options/general.py [hunk 3]
```diff
@@ -46,6 +46,7 @@ class GeneralOptionsPage(OptionsPage):
         config.TextOption("persist", "oauth_access_token", ""),
         config.IntOption("persist", "oauth_access_token_expires", 0),
         config.TextOption("persist", "oauth_username", ""),
+        config.IntOption("setting", "file_import_threshold", 5000),
```
Comment (zas) [inline picard/ui/options/general.py]: Please rename the option to `file_import_warning_threshold`

File: picard/tagger.py [hunk 4]
```diff
@@ -358,17 +359,45 @@ def add_files(self, filenames, target=None):
...
+    def check_load(self, new_files):
+        #Load large amounts of files
+        num_files = len(new_files)
+        if num_files > config.setting["file_import_threshold"]:
+            QMessageBox = QtGui.QMessageBox
+
+            msg = QMessageBox()
+            msg.setIcon(QMessageBox.Warning)
+            msg.setWindowModality(QtCore.Qt.WindowModal)
+            msg.setWindowTitle(_(u"Large File Load"))
```
Comment (zas) [inline picard/tagger.py]: Title of the dialog should indicate it is a warning, "Large File Load" looks unrelated.
What about "Warning: too many files at once" ? or something like that.

File: picard/tagger.py [hunk 5]
```diff
@@ -358,17 +359,45 @@ def add_files(self, filenames, target=None):
...
+            msg.setText(
+                ungettext(
+                    u"Are you sure you want to load %d file? Picard may run very slowly" %num_files,
+                    u"Are you sure you want to load %d files? Picard may run very slowly" %num_files,
```
Comment (zas) [inline picard/tagger.py]: Add a space after `%`

File: picard/tagger.py [hunk 6]
```diff
@@ -358,17 +359,45 @@ def add_files(self, filenames, target=None):
...
+                    num_files))
+            cancel = msg.addButton(QMessageBox.Cancel)
+            msg.setDefaultButton(cancel)
+            msg.addButton(_(u"&Load"), QMessageBox.YesRole)
```
Comment (zas) [inline picard/tagger.py]: I would just use Yes/No button here, as the question is "Are you sure...."

File: picard/tagger.py [hunk 7]
```diff
@@ -358,17 +359,45 @@ def add_files(self, filenames, target=None):
...
             if target is None or target is self.unmatched_files:
                 self.unmatched_files.add_files(new_files)
                 target = None
             for file in new_files:
                 file.load(partial(self._file_loaded, target=target))
 
+    def check_load(self, new_files):
+        #Load large amounts of files
+        num_files = len(new_files)
+        if num_files > config.setting["file_import_threshold"]:
```
Comment (zas) [inline picard/tagger.py]: You need to change option name here too.

File: picard/ui/options/general.py [hunk 8]
```diff
@@ -62,12 +63,14 @@ def load(self):
         self.ui.server_port.setValue(config.setting["server_port"])
         self.ui.analyze_new_files.setChecked(config.setting["analyze_new_files"])
         self.ui.ignore_file_mbids.setChecked(config.setting["ignore_file_mbids"])
+        self.ui.file_import_threshold.setValue(config.setting["file_import_threshold"])
```
Comment (zas) [inline picard/ui/options/general.py]: s/file_import_threshold/file_import_warning_threshold/

File: picard/ui/options/general.py [hunk 9]
```diff
@@ -62,12 +63,14 @@ def load(self):
...
 
     def save(self):
         config.setting["server_host"] = unicode(self.ui.server_host.currentText()).strip()
         config.setting["server_port"] = self.ui.server_port.value()
         config.setting["analyze_new_files"] = self.ui.analyze_new_files.isChecked()
         config.setting["ignore_file_mbids"] = self.ui.ignore_file_mbids.isChecked()
+        config.setting["file_import_threshold"] = self.ui.file_import_threshold.value()
```
Comment (zas) [inline picard/ui/options/general.py]: s/file_import_threshold/file_import_warning_threshold/

File: ui/options_general.ui [hunk 10]
```diff
@@ -133,17 +133,55 @@
...
      </property>
-     <property name="sizeHint" stdset="0">
-      <size>
-       <width>181</width>
-       <height>21</height>
-      </size>
-     </property>
-    </spacer>
+     <item row="0" column="1">
+       <widget class="QSpinBox" name="file_import_threshold">
```
Comment (zas) [inline ui/options_general.ui]: s/file_import_threshold/file_import_warning_threshold/

File: picard/ui/ui_options_general.py [hunk 11]
```diff
@@ -98,4 +118,5 @@ def retranslateUi(self, GeneralOptionsPage):
         self.groupBox_2.setTitle(_("General"))
         self.analyze_new_files.setText(_("Automatically scan all new files"))
         self.ignore_file_mbids.setText(_("Ignore MBIDs when loading new files"))
-
+        self.import_settings.setTitle(_("Importing"))
+        self.label_import.setText(_("Warn if user tries to load more than this number of files at once:"))
```
Comment (zas) [inline picard/ui/ui_options_general.py]: The change should have been done in `ui/options_general.ui`

File: ui/options_general.ui [hunk 12]
```diff
@@ -133,17 +133,55 @@
...
+      <item row="0" column="0">
+       <widget class="QLabel" name="label_import">
+        <property name="sizePolicy">
+         <sizepolicy hsizetype="Expanding" vsizetype="Fixed">
+          <horstretch>0</horstretch>
+          <verstretch>0</verstretch>
+         </sizepolicy>
+        </property>
+        <property name="text">
+         <string>Maximum number of files to import without error:</string>
```
Comment (zas) [inline ui/options_general.ui]: Change here

**Comment by zas** [comment]
I found a major issue with this PR: when one drops a top level directory, the warning dialog will only consider each subdirectory, and therefore don't count all files to be imported and even worse, shows the dialog for each subdir....
For example, let's say the limit is set to 5, and i have 100 directories of 6 files, and drop the parent directory of those 100, it will not show a warning because 600 > 5 (expected) but 100 warnings for 6 > 5 (unexpected).

**Comment by zas** [comment]
Replaced by https://github.com/metabrainz/picard/pull/613
